### DISCUSSION TITLE
Can't connect to Elfin EW11a

### DISCUSSION THREAD

**Discussion by mschae** [nathanmarlor/foxess_modbus#273] [Support]
## Version of the custom_component

v1.5.2

## Configuration

```yaml
Add your logs here.
```

## Describe the bug

When attempting to connect to the adapter for the first time I get the following error message:

> Unable to connect to your adapter. Ensure that the adapter is properly configured (see the setup link above) and the connection details are correct, then try again. Details: Modbus Error: [Connection] ModbusTcpClient(10.112.1.214:502): Connection unexpectedly closed 1.0967254638671875e-05 seconds into read of 8 bytes without response from unit before it closed connection.

The adapter seems to receive the "init" frames and it looks like it can talk to the inverter but closes the connection prematurely: 
<img width="747" alt="Bildschirm­foto 2023-05-23 um 13 30 26" src="https://github.com/nathanmarlor/foxess_modbus/assets/473241/de16f837-8966-4f68-95f5-d255bb7c0d82">

## Debug log

```text

Logger: custom_components.foxess_modbus.modbus_controller
Source: custom_components/foxess_modbus/modbus_client.py:160 
Integration: FoxESS - Modbus ([documentation](https://github.com/nathanmarlor/foxess_modbus), [issues](https://github.com/nathanmarlor/foxess_modbus/issues)) 
First occurred: 12:05:17 (35 occurrences) 
Last logged: 13:33:27

Autodetect: failed to connect to (tcp://10.112.1.214:502)
Autodetect: failed to connect to (tcp://10.222.0.213:502)
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/pymodbus/client/tcp.py", line 309, in recv
    ready = select.select([self.socket], [], [], end - time_)
ValueError: filedescriptor out of range in select()

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/foxess_modbus/modbus_controller.py", line 332, in autodetect
    await client.read_registers(
  File "/config/custom_components/foxess_modbus/modbus_client.py", line 73, in read_registers
    response = await self._async_pymodbus_call(
  File "/config/custom_components/foxess_modbus/modbus_client.py", line 160, in _async_pymodbus_call
    result = await self._hass.async_add_executor_job(call, *args)
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/site-packages/pymodbus/client/mixin.py", line 100, in read_holding_registers
    return self.execute(
  File "/usr/local/lib/python3.10/site-packages/pymodbus/client/base.py", line 192, in execute
    return self.transaction.execute(request)
  File "/usr/local/lib/python3.10/site-packages/pymodbus/transaction.py", line 159, in execute
    response, last_exception = self._transact(
  File "/usr/local/lib/python3.10/site-packages/pymodbus/transaction.py", line 306, in _transact
    result = self._recv(response_length, full)
  File "/usr/local/lib/python3.10/site-packages/pymodbus/transaction.py", line 341, in _recv
    read_min = self.client.framer.recvPacket(min_size)
  File "/usr/local/lib/python3.10/site-packages/pymodbus/framer/__init__.py", line 53, in recvPacket
    return self.client.recv(size)  # pylint: disable=no-member
  File "/usr/local/lib/python3.10/site-packages/pymodbus/client/tcp.py", line 311, in recv
    return self._handle_abrupt_socket_close(size, data, time.time() - time_)
  File "/usr/local/lib/python3.10/site-packages/pymodbus/client/tcp.py", line 363, in _handle_abrupt_socket_close
    raise ConnectionException(msg)
pymodbus.exceptions.ConnectionException: Modbus Error: [Connection] ModbusTcpClient(10.112.1.214:502): Connection unexpectedly closed 3.62396240234375e-05 seconds into read of 8 bytes without response from unit before it closed connection
```

**Comment by canton7** [comment]
First off, have you followed the configuration in https://github.com/nathanmarlor/foxess_modbus/wiki/Elfin-EW11?

**Reply by mschae** [reply]
I have. Let me post screenshots:
<img width="750" alt="Bildschirm­foto 2023-05-23 um 13 57 02" src="https://github.com/nathanmarlor/foxess_modbus/assets/473241/e65b7e32-4ca4-4e2f-a887-065a61941200">

<img width="738" alt="image" src="https://github.com/nathanmarlor/foxess_modbus/assets/473241/5315646f-93d8-47b7-bf9b-61b8d8af54a8">

**Reply by mschae** [reply]
FWIW it looks similar to this one: https://github.com/nathanmarlor/foxess_modbus/discussions/227

**Reply by canton7** [reply]
Yeah I wondered that. I was going to post the link, but I noticed that we suggested checking the wiring and then OP went quiet, which implies that wiring was the problem. But from your status pages it does seem that you're getting a response back, which suggests it isn't a wiring problem.

Looking at the sizes, we're attempting to read 15 registers during that initial connection. They're 2 bytes each, + 5 bytes, so we're expecting 35 bytes back, unless I've mis-calculated something. That matches the 2 frames of 70 bytes between them.

The request should be 12 bytes over TCP, so that again matches the 2 frames of 24 bytes. When sent over the serial side that's 8 bytes each, which matches the 16.

So it does look like your initial diagnosis is right: the EW11 receives the response from the inverter correctly, but then it (or something else) terminates the connection before sending them back to home assistant.

I've skimmed the pymodbus code, and it looks like it should only throw that error if the socket has been closed by the remote side.

At the moment I'm suspecting a firewall or something jumping into the connection and terminating it. Do you have any fancy network setup which might cause that? Are you familiar enough with Wireshark/tcpdump to have a look at the packets going into / out of your home assistant machine, see what's terminating the connection?

**Reply by canton7** [reply]
It could also be an IP address conflict I guess, although it doesn't quite match the symptoms: I'd expect a timeout, rather than a connection that's been deliberately closed. Are you sure that nothing else on your network is using the EW11's IP address?

**Reply by mschae** [reply]
Sorry for the late reply and thanks for your awesome support up to this point. Is there anything I can use from the CLI to test the modbus via tcp connection?

I can rule address issues out: I've connected the EW11 to two different networks and tried two different IP addresses. Both yielded the same result. Addresses are strictly DHCP in my network and sind I can access the webpage just fine I'd rule out network issues.

**Reply by mschae** [reply]
What's weird is that the TCP server is reporting 0 frames sent back. Is that normal? FWIW I did install the latest firmware with no different result.

**Reply by mschae** [reply]
I can try to run a tcpdump, I use different VLANs on OpenWRT so I could go as far as to capture packets on the router itself. Since it's TCP it should work fine tho: The HA instance is in a VLAN which has full access to the VLAN that the EW11 is in and as written earlier I even put them on the same network/VLAN to rule out anything there...

**Reply by calum-mcfarlane** [reply]
Have a look at the Troubleshooting section of [this wiki entry](https://github.com/nathanmarlor/foxess_modbus/wiki/USR-TCP232-410s) (scroll to the bottom).  There's lots of modbus related software out there but Modpoll is one of the simplest to use for testing basic functionality.

**Reply by canton7** [reply]
You can use modpoll or similar, [see here](https://github.com/StealthChesnut/HA-FoxESS-Modbus/wiki/USR-W610-Setup-Guide#troubleshooting-usr) (I really need to do a proper write-up of how to use it). Use e.g. `c:\path-to-modpoll\modpoll.exe -m tcp -p 502 -a 247 -t 3 -0 -1 -c 15 -r 30000 x.x.x.x.x`.

> What's weird is that the TCP server is reporting 0 frames sent back. Is that normal? FWIW I did install the latest firmware with no different result.

My guess is still that a firewall or something is jumping in and terminating the TCP connection -- the speed it which it closes indicates that something is sending a FIN, and that would explain why the EW11 doesn't try and send a response over TCP. But that's a guess -- it might be that the EW11 is closing the connection itself for some reason.

**Reply by canton7** [reply]
I'd run tcpdump on the HA machine itself if you can to start with -- let's see what it's seeing, and whether it does receive a FIN packet. If it does, tcpdump on the router might help narrow down whether that's coming from the EW11 or something else (like the router).

**Reply by mschae** [reply]
```
# tcpdump -i br-lan host 10.222.0.6 and tcp port 502
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on br-lan, link-type EN10MB (Ethernet), capture size 262144 bytes

16:23:58.394391 IP unifi.43616 > 10.112.1.214.502: Flags [S], seq 1804199454, win 29200, options [mss 1460,sackOK,TS val 1042521124 ecr 0,nop,wscale 7], length 0
16:23:58.394456 IP unifi.43616 > 10.112.1.214.502: Flags [S], seq 1804199454, win 29200, options [mss 1460,sackOK,TS val 1042521124 ecr 0,nop,wscale 7], length 0
16:23:58.396673 IP 10.112.1.214.502 > unifi.43616: Flags [S.], seq 19074, ack 1804199455, win 4380, options [mss 1460], length 0
16:23:58.396723 IP 10.112.1.214.502 > unifi.43616: Flags [S.], seq 19074, ack 1804199455, win 4380, options [mss 1460], length 0
16:23:58.397015 IP unifi.43616 > 10.112.1.214.502: Flags [.], ack 1, win 29200, length 0
16:23:58.397038 IP unifi.43616 > 10.112.1.214.502: Flags [.], ack 1, win 29200, length 0
16:23:58.397244 IP unifi.43616 > 10.112.1.214.502: Flags [P.], seq 1:13, ack 1, win 29200, length 12
16:23:58.397265 IP unifi.43616 > 10.112.1.214.502: Flags [P.], seq 1:13, ack 1, win 29200, length 12
16:23:58.397244 IP unifi.43616 > 10.112.1.214.502: Flags [F.], seq 13, ack 1, win 29200, length 0
16:23:58.397288 IP unifi.43616 > 10.112.1.214.502: Flags [F.], seq 13, ack 1, win 29200, length 0
16:23:58.399621 IP 10.112.1.214.502 > unifi.43616: Flags [.], ack 14, win 4367, length 0
16:23:58.399645 IP 10.112.1.214.502 > unifi.43616: Flags [.], ack 14, win 4367, length 0
16:23:58.401886 IP 10.112.1.214.502 > unifi.43616: Flags [F.], seq 1, ack 14, win 4367, length 0
16:23:58.401905 IP 10.112.1.214.502 > unifi.43616: Flags [F.], seq 1, ack 14, win 4367, length 0
16:23:58.402162 IP unifi.43616 > 10.112.1.214.502: Flags [.], ack 2, win 29200, length 0
16:23:58.402181 IP unifi.43616 > 10.112.1.214.502: Flags [.], ack 2, win 29200, length 0
```

Here's another capture as pcap: 
[output.pcap.zip](https://github.com/nathanmarlor/foxess_modbus/files/11545203/output.pcap.zip)

**Reply by mschae** [reply]
Sorry saw your answer too late, that's from my router

**Reply by mschae** [reply]
Here is what my home assistant host sees:

```
$ sudo tcpdump -i enp2s0f0 tcp port 502
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on enp2s0f0, link-type EN10MB (Ethernet), capture size 262144 bytes
16:35:38.072380 IP unifi.lan.43640 > EW11.lan.502: Flags [S], seq 145209336, win 29200, options [mss 1460,sackOK,TS val 1042696053 ecr 0,nop,wscale 7], length 0
16:35:38.074738 IP EW11.lan.502 > unifi.lan.43640: Flags [S.], seq 51389, ack 145209337, win 4380, options [mss 1460], length 0
16:35:38.074801 IP unifi.lan.43640 > EW11.lan.502: Flags [.], ack 1, win 29200, length 0
16:35:38.089780 IP unifi.lan.43640 > EW11.lan.502: Flags [P.], seq 1:13, ack 1, win 29200, length 12
16:35:38.089862 IP unifi.lan.43640 > EW11.lan.502: Flags [F.], seq 13, ack 1, win 29200, length 0
16:35:38.092291 IP EW11.lan.502 > unifi.lan.43640: Flags [.], ack 14, win 4367, length 0
16:35:38.094300 IP EW11.lan.502 > unifi.lan.43640: Flags [F.], seq 1, ack 14, win 4367, length 0
16:35:38.094341 IP unifi.lan.43640 > EW11.lan.502: Flags [.], ack 2, win 29200, length 0
```

**Reply by canton7** [reply]
What IP is what device? 10.112.1.214.502 is the EW11 presumably, and the unifi will be the router.

I'll look at the pcap at some point when I'm at home -- can't have wireshark on a mobile.

It looks like the router sends the modbus request (the packet with length 12), and then immediately sends a FIN.

However, I don't understand why everything is duplicated. Sure that's an artifact of how it was captured -- maybe we're capturing traffic as it crosses the bridge or something, those duplicate packets can't actually be being sent to the EW11.

**Reply by canton7** [reply]
Is HA running on unifi.lan?

**Reply by canton7** [reply]
Googling around, the problem might be that you're out of fds: pymodbus is calling `select` at that point, which fails if you try and give it an fd > 1024. It'd be interesting to see `lsof | wc -l` on the HA machine.

**Reply by canton7** [reply]
Yep, I'm pretty sure that's the problem. Looks like pymodbus has seen this error before, but mis-diagnosed it as the remote side terminating the connection early... Yay for python libraries 🙄 

Looks like we can hack our way around it by moving that specific bit of functionality to use poll.

**Reply by canton7** [reply]
For now, try restarting -- that might clear out some FDs (as a temporary solution). Particularly if you're running HA on a unifi, I'd expect there to be a lot of active connections.

**Comment by canton7** [comment]
@mschae Can you give the latest pre-release v1.5.3b1 a spin?

I've hacked pymodbus to use poll() rather than select(). If I'm right in my diagnosis, that should fix the issue

**Reply by canton7** [reply]
@mschae Have you had a chance to try this out, by any chance?
